hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java

```diff
@@ -0,0 +1,341 @@
+    ) {
+        super(name);
+        this.indexName = indexName;
+        this.partitionIdSet = partitionIdSet;
+        this.pointers = pointers;
+        this.sizeHint = sizeHint;
+    }
+
+    @Override
+    protected void runInternal() {
+        partitionService = getNodeEngine().getPartitionService();
+        int partitionCount = partitionService.getPartitionCount();
+        Address currentAddress = getNodeEngine().getLocalMember().getAddress();
+
+        Indexes indexes = mapContainer.getIndexes();
+        if (indexes == null) {
+            throw new QueryException("Cannot use the index \"" + indexName + "\" of the IMap \""
+                            + mapContainer.getName() + "\" because it is not global "
+                            + "(make sure the property \"" + ClusterProperty.GLOBAL_HD_INDEX_ENABLED
+                            + "\" is set to \"true\")");
+        }
+
+        InternalIndex index = indexes.getIndex(indexName);
+        if (index == null) {
+            throw new QueryException("Index name \"" + indexName + "\" does not exist");
+        }
+
+        if (!allIndexed(index, partitionIdSet)) {
+            throw new QueryException("Some of the partitions are not indexed in \"" + indexName + "\"");
+        }
+
+        int startMigrationTimestamp = getMigrationTimestamp();
+
+        ResultInternal result;
+        switch (index.getConfig().getType()) {
+            case HASH:
+                result = runInternalHash(index, pointers, partitionIdSet, sizeHint);
+                break;
+            case SORTED:
+                result = runInternalSorted(index, pointers, partitionIdSet, sizeHint);
+                break;
+            case BITMAP:
+                throw new UnsupportedOperationException("BITMAP index scan is not implemented");
+            default:
+                throw new UnsupportedOperationException(
+                        "Unknown index type: \"" + index.getConfig().getType().name() + "\"");
+        }
+
+        List<QueryableEntry> entries = result.getResult();
+        IndexIterationPointer[] newPointers = result.getPointers();
+
+        int endMigrationTimestamp = getMigrationTimestamp();
+
+        // In case of migration, discard results, declare migrated partitions
+        // and return old pointers
+        if (endMigrationTimestamp != startMigrationTimestamp) {
+            List<Integer> currentPartitions = partitionService.getMemberPartitions(currentAddress);
+            PartitionIdSet currentPartitionIdSet = new PartitionIdSet(partitionCount, currentPartitions);
+
+            Set<Integer> migratedPartitionIds =
+                    partitionIdSet
+                            .stream()
+                            .filter(id -> !currentPartitionIdSet.contains(id))
+                            .collect(Collectors.toSet());
+
+            response = new MapFetchIndexOperationResult(Collections.emptyList(), migratedPartitionIds, pointers);
+            return;
```

Comment:
We don't have to return the list of missing partitions to the caller. It can and even has do it itself - for example it has to figure out who's the new owner of the missing partitions. We can just throw an exc. here